sentry-android-core/src/main/java/io/sentry/android/core/ANRWatchDog.java

```diff
@@ -100,48 +126,49 @@ public void run() {
-
-        if (am != null) {
-          List<ActivityManager.ProcessErrorStateInfo> processesInErrorState = null;
-          try {
-            // It can throw RuntimeException or OutOfMemoryError
-            processesInErrorState = am.getProcessesInErrorState();
-          } catch (Throwable e) {
-            logger.log(
-                SentryLevel.ERROR, "Error getting ActivityManager#getProcessesInErrorState.", e);
-          }
-          // if list is null, there's no process in ANR state.
-          if (processesInErrorState == null) {
-            continue;
-          }
-          boolean isAnr = false;
-          for (ActivityManager.ProcessErrorStateInfo item : processesInErrorState) {
-            if (item.condition == NOT_RESPONDING) {
-              isAnr = true;
-              break;
-            }
-          }
-          if (!isAnr) {
-            continue;
-          }
-        }
+        if (isProcessNotResponding() && reported.compareAndSet(false, true)) {
+          final String message =
+              "Application Not Responding for at least " + timeoutIntervalMillis + " ms.";
 
-        logger.log(SentryLevel.INFO, "Raising ANR");
-        final String message =
-            "Application Not Responding for at least " + timeoutIntervalMillis + " ms.";
+          final ApplicationNotResponding error =
+              new ApplicationNotResponding(message, uiHandler.getThread());
+          anrListener.onAppNotResponding(error);
+        }
+      }
+    }
+  }
 
-        final ApplicationNotResponding error =
-            new ApplicationNotResponding(message, uiHandler.getThread());
-        anrListener.onAppNotResponding(error);
-        interval = timeoutIntervalMillis;
+  private boolean isProcessNotResponding() {
+    // we only raise an ANR event if the process is in ANR state.
+    // if ActivityManager is not available, we'll still be able to send ANRs
+    final ActivityManager am = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
 
-        reported.set(true);
+    if (am != null) {
+      List<ActivityManager.ProcessErrorStateInfo> processesInErrorState = null;
+      try {
+        // It can throw RuntimeException or OutOfMemoryError
+        processesInErrorState = am.getProcessesInErrorState();
+      } catch (Throwable e) {
+        logger.log(SentryLevel.ERROR, "Error getting ActivityManager#getProcessesInErrorState.", e);
+      }
+      // if list is null, there's no process in ANR state.
+      if (processesInErrorState != null) {
+        for (ActivityManager.ProcessErrorStateInfo item : processesInErrorState) {
+          if (item.condition == NOT_RESPONDING) {
+            return true;
+          }
+        }
       }
+      // when list is empty, or there's no element NOT_RESPONDING, we can assume the app is not
+      // blocked
+      return false;
     }
+    return true;
+  }
+
+  @FunctionalInterface
+  interface TimeProvider {
```

Comment:
`l`: we also have `ICurrentDateProvider/AndroidCurrentDateProvider`, would it make sense to reuse it here instead?